### PR TITLE
More reliable log handling in Keeper

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Keeper fix: don't remove the entire log if some entries can't be read. This can happen if the last uncommitted data was not completely written to the disk.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 998dc12c98484c38e05999415a6cef565e1f5be0 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56670/998dc12c98484c38e05999415a6cef565e1f5be0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6849790336">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56670/998dc12c98484c38e05999415a6cef565e1f5be0/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Coordination/CoordinationSettings.cpp [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #include <Coordination/CoordinationSettings.h>
+#include "Common/Exception.h"
```
Comment (alesapin) [inline src/Coordination/CoordinationSettings.cpp:2]:
```suggestion
#include <Common/Exception.h>
```

File: src/Coordination/Changelog.cpp [hunk 2]
```diff
@@ -813,37 +823,42 @@ void Changelog::readChangelogAndInitWriter(uint64_t last_commited_log_index, uin
...
+        if (last_log_read_result->last_read_index == 0) /// If it's broken or empty log then remove it
         {
-            LOG_INFO(log, "Removing chagelog {} because it's empty or read finished with error", description->path);
+            LOG_INFO(log, "Removing chagelog {} because it's empty", description->path);
+            remove_invalid_logs();
             description->disk->removeFile(description->path);
             existing_changelogs.erase(last_log_read_result->log_start_index);
             std::erase_if(logs, [last_log_read_result](const auto & item) { return item.first >= last_log_read_result->log_start_index; });
         }
+        else if (last_log_read_result->error)
```
Comment (alesapin) [inline src/Coordination/Changelog.cpp:847]: The main fix is here, right? Now if we have error we will not remove broken log.

**Review by alesapin — APPROVED** [review]

File: src/Coordination/Changelog.cpp (hunk 2 above)
Comment (antonio2368) [reply src/Coordination/Changelog.cpp:847]: yes, and we continue reading if we detect error
https://github.com/ClickHouse/ClickHouse/pull/56670/files#diff-f1188c18e99d4fe945c6a979a345bb79c30a372c2c3b39524f312f70649a0648L766

gaps are still caught with the check that the differences between logs cannot be more than 1

**Pull request merged**